Add data read

oruebel opened the pull request:
This PR is to try and implement the proposed approach for data read from #83 to see if this approach is viable. This PR is experimental right now and should not be merged. 



## 2. Proposed Implementation for reading data arrays

### ``BaseReadData``
- [x] Create a new `ReadDatasetWrapper` and `ReadAttributeWrapper` classes for reading data. This is modified from the proposal, which suggested a single  `BaseReadData` class for reading any array (datasets, attributes) from a file. 
- [x] Support conversion to boost multi-dimensional array for convenience 
- **Note** I did not update `BaseRecordingData` to inherit from `ReadDataWrapper` because the two are not compatible right now. `BaseReadData` uses the io and the path to access the data, whereas `BaseRecordingData` leaves that choice to the I/O backend and stores the references to the dataset. We may or may not want to change this. 

### ``BaseIO``
- **Note** I did not add abstract methods for lazy reading objects from a file to `BaseIO` (or more accurately I remove them) because: 1) I wanted to use the shared_ptr to the I/O rather than a raw pointer, which I can't get from BaseIO, and 2) with the ReadDatasetWrapper this is more approbriately done in the Container class directly.
- [X] Add pure virtual method to allow us to get the storage object type (Group, Dataset, Attribute) for a given path
- [X] Add pure virtual methods to read data values from a Dataset or Attribute that the BaseReadData can call for read

### ``HDF5IO``
- **Note** In contrast to the proposal, I did not implement  specific version of the `BaseReadData` for `HDF5` but left read logic to `HDF5IO` itself so that the `ReadDatasetWrapper`  can remain generic. To make this more manageable, I defined `ReadAttributeWrapper` separately
- [X] Implement the methods for reading data values from a Dataset or Attribute that the ``HDF5ReadDataSet`` and ``HDF5ReadAttribute`` wrappers can call can call for read
- [X] Implement the ``getObjectType`` method for getting the storage object type (Group, Dataset, Attribute) for a given path


### ``Container``

- [X] Store the `io` object on the `Container` so that we can call ``io->readDataset`` and ``io->readAttribute`` in the read methods

### NWB types: ``TimeSeries``, ``ElectricalSeries`` etc.
- [x] Remove storage of properties from the ``Container`` classes and replace them with access methods that return `BaseReadData` objects instead. This allows for reading in both read and write mode and avoids keeping data in memory that we have already written to disk. For example, in `TimeSeries`, these variables would need to change to properties: https://github.com/NeurodataWithoutBorders/aqnwb/blob/e873d95a1eb998dd97894b665205a7ab8608e872/src/nwb/base/TimeSeries.hpp#L91-L140

## 3. Proposed implementation for reading whole ``Containers`` (e.g., to read an ``ElectricalSeries``)

- Add access methods on the respective ``Container`` that owns the respective objects, e.g., ``NWBFile`` owning ``ElectricalSeries`` objects to retrieve the object
- [X] Add abstract factory method (that is templated on the return type) to ``Container``  to create an instance of the specific ``Container`` type using only the `io` and `path` for the `Container` as input. The specific ``Container`` classes, such as ``TimeSeries`` will then need to implement a corresponding constructor that uses ``io`` and ``path`` as input. 

#### Step 1: Define the Template Factory Method in ``Container`` 

```
class Container {
public:
   
    template <typename T>
    static std::unique_ptr<T> create(const BaseIO& io, const std::string& path) {
        static_assert(std::is_base_of<Container, T>::value, "T must be a derived class of Container");
        return std::unique_ptr<T>(new T(path, io));
    }
};
```

#### Step 2: Implement the constructors on the specific ``Container`` classes (e.g., ``TimeSeries``)

- [X] Add the necessary constructor

```
class TimeSeries : public Container {
public:
    TimeSeries(const std::string& path, const BaseIO& io) {
        // Implementation of TimeSeries constructor
    }
};
```

## 4. Proposed implementation for reading untyped groups (e.g., `/acquisition`)

I'm not sure we'll need do this, since a group by itself does not define data. To access the contents we could define access methods on the parent ``Container`` class (e.g., ``NWBFile``) that owns the untyped group to access its contents. 


# TODO
- [X] Implement method to construct Container for read from NWBFile  (See #91)
- [X] Check all Container classes (e.g., ElectrodeGroup etc.) to make sure we move all the arguments (other than path and io) to the initialize method. Done in #91
- [X] Check if we can fix data type to retrieve type `ReadDatasetWrapper`, `ReadAttributeWrapper`  (See #91)
- [X] Check how we can allow the user query the data type to use for an object (See #91)
- [X] Clean up storage of properties that should be read (See #91)
- [X] Debug reading of attributes
- [X] Test reading of string datasets and attributes
- [X] Add more detailed docs for reading
- [x] Move code from functions defined in the header files (mainly for quick testing) to the cpp files 
- [x] Add common base class for ``ReadDatasetWrapper`` and ``ReadAttributeWrapper`` (I think this could be useful)
- [x] Add support for storing the native data type on the DataBlockGeneric to make it easier for the user to inspect it
- [x] Move BaseIO and HDF5 to a new IO module Fix #88 
- [x] Refactor BaseIO to separate certain classes into their own hpp/cpp files

# Next steps
- [x] Check that workflows are running the correct code version. The build in the test fails with /Users/runner/work/aqnwb/aqnwb/src/nwb/ecephys/ElectricalSeries.cpp:61:25: error: use of undeclared identifier 'BaseDataType'; did you mean 'IO::BaseDataType'? m_io->createAttribute(BaseDataType::I32, But that is not what in line 61 in this PR https://github.com/NeurodataWithoutBorders/aqnwb/actions/runs/11023552242/job/30615439516?pr=85#step:6:249
- [x] Once the workflow issue is resolved, I propose we merge #91 with this PR to create a single PR
- [x] See #91 for open ToDo items that should be addressed before merge
- [x] Merge PR #123 
- [x] Merge PR #128 
- [x] Merge PR #124 
- [x] Merge PR #118  
- [x] Add  unit tests for public ``HDF5IO``-specific functions (i.e., not inherited from ``BaseIO``: 
    - [x] ``HDF5IO::getH5ObjectType``. Implemented in #118  
    - [x] ``HDF5IO::getNativeType``. Implemented in #118 
    - [x] ``HDF5IO::getH5Type``. Implemented in #118 
- [x] Add unit tests for new  public``HDF5IO`` functions (which are core functions new in``BaseIO``). While some of these are partially covered by the ``DEFINE_FIELDS`` read tests, it will be useful to more fully test these separately  
    - [x] ``HDF5IO::readDataset`` (incl. with start, count, stride, and block parameters set). This is to test that reading subsets of larger arrays works as expected. Implemented in #118 
    - [x] ``HDF5IO::open(FileMode mode)`` test that file modes are being set correctly.   Implemented in #118 
    - [x] ``HDF5IO::readAttribute``.  Implemented in #118 
    - [x] ``HDF5IO::getStorageObjectType``.  Implemented in #118 
    - [x] ``HDF5IO::getGroupObjects``.  Implemented in #118 
    - [x] ``HDF5IO::attributeExists``.  Implemented in #118 
    - [x] ``HDF5IO::objectExists`` . Implemented in #118 
- [x] Update ``HDF5IO::readDataset`` to make sure it supports all supported types of BaseDataType, in partiuclar also strings and update the``TEST_CASE("HDF5IO; read dataset", "[hdf5io]")`` and ``TEST_CASE("HDF5IO; read dataset subset", "[hdf5io]")`` (See #118 )
- [x] Fix  nwb-inspector violations. (``cd build/dev/tests/data`` and run ``nwbinspector . --threshold BEST_PRACTICE_VIOLATION``) see #125 
- [x] Resolve open conversations: 
    - [x] Resolve comment on updating workflows to run all PRs https://github.com/NeurodataWithoutBorders/aqnwb/pull/85/files#r1779068283
    - [x] Make `DEFINE_FIELD` create virtual functions and resolve this comment https://github.com/NeurodataWithoutBorders/aqnwb/pull/85/files#r1894223365  **NOTE:** I created #117 for this and added a note to the docs to example this. I think it is fine to address this later. 
    - [x] Fix failing Windows build due to ``'boost/multi_array.hpp': No such file or directory``
    - [X] Resolve discussion on how to potentially avoid the need for `DataBlockGeneric` and use only `DataBlock` https://github.com/NeurodataWithoutBorders/aqnwb/pull/85/files#r1773994249  (see #132)


# Items moved to new issues
* #115
* #114 
* #113
* #127
* #129 and also related to #107
* #130
* #131
* #132
* #133
* #134

oruebel commented:
@stephprince when you get a chance ,could you please do a first code review of this PR to make sure this is heading in the right direction. I now have a first outline of one possible solution for how we might implement read. There is still a lot more work to be done before this PR is ready, but it would be useful if you could take a look before I go any further with this approach.

oruebel commented:
@stephprince I just added a documentation page as well, which hopefully helps explain the current proposed design for read so we can review and discuss.

oruebel commented:
@stephprince I made a couple of adjustments since we talked about the PR to deal with typing of the ReadDatasetWrapper etc. for reading data. Specifically:
- I merge in the latest changes from the `main` branch from the refactoring 
- I modified `ReadDatasetWrapper` and `ReadAttributeWrapper` to be templated by the type using `std::any` (i.e., unknown) as the default. In this way we can fix the data type in the API for objects where we know the type from the schema
- I modified `DataBlockGeneric` (and `DataBlock`) to also store the `std::type_index`. The `HDF5IO` backend already determines the data type, so we can just save it. In this way the user has access to the the data type at runtime and can implement appropriate logic if necessary.

stephprince commented on src/io/hdf5/HDF5IO.cpp:
```diff
@@ -82,7 +83,305 @@ Status HDF5IO::flush()
+std::unique_ptr<H5::Attribute> HDF5IO::getAttribute(const std::string& path)
+{
+  // Split the path to get the parent object and the attribute name
+  size_t pos = path.find_last_of('/');
+  if (pos == std::string::npos) {
+    return nullptr;
+  }
+
+  std::string parentPath = path.substr(0, pos);
+  std::string attrName = path.substr(pos + 1);
```
these manipulations as well as the `mergePath` functions I think are good reasons to switch to using the `std::filesystem::path representation`?

oruebel replied:
The issue with "std::filesystem::path" is that is will use "/" or "\" depending on the system it is running on (Windows vs Unix), but HDF5 always uses "/"

stephprince commented on src/io/hdf5/HDF5IO.cpp:
```diff
@@ -82,7 +83,305 @@ Status HDF5IO::flush()
+  std::string parentPath = path.substr(0, pos);
+  std::string attrName = path.substr(pos + 1);
+
+  try {
```
could use `getH5ObjectType` here instead of trying to open object as group or a dataset?

stephprince commented on docs/pages/userdocs/read.dox:
```diff
@@ -0,0 +1,360 @@
+ *
+ * \snippet tests/examples/test_ecephys_data_read.cpp example_read_get_datablock_snippet
+ *
+ * The data is here represented as a \ref AQNWB::IO::DataBlock "DataBlock", which stores the data as 1-dimensionsal
```
```suggestion
 * The data is here represented as a \ref AQNWB::IO::DataBlock "DataBlock", which stores the data as 1-dimensional
```

stephprince reviewed (commented):
Finally got a chance to look through this and added some comments / questions! A couple of comments In addition to what I added in code:
1.  I think we want to add tests for reading in string datasets, attributes, and multidimensional datasets. There were a couple of sections in the docs examples maybe working on testing these things that were commented out, not sure if those were WIP or resolved elsewhere.
2. How do links/references work on read?

I'm partway through looking through the follow up PR, will ping you there when I finish.

oruebel commented:
> 1. think we want to add tests for reading in string datasets, attributes, and multidimensional datasets. There were a couple of sections in the docs examples maybe working on testing these things that were commented out, not sure if those were WIP or resolved elsewhere.

Totally agree. This PR + #91 together outline how read can work, but there are still a lot of tests that need to be added to make sure everything is working as expected. I think in particular on the ``HDF5IO`` there are probably still some gremlins hiding. This is part of the todo list in #91.

oruebel commented:
> 2\. How do links/references work on read?

Links should be handled transparently by HDF5 so I think read should work in the same way as for groups and datasets. I have not yet addressed the case of object references that are stored as values of attributes and datasets. I think this will require additional logic in the `ReadDataWrapper` and `HDF5IO`.

oruebel commented on .github/workflows/codespell.yml:
```diff
@@ -6,8 +6,8 @@ on:
   pull_request:
-    branches:
-    - main
+    #branches:
```
The reason I had modfied this is because it would'nt trigger workflows if they didn't target the main branch. I think it would be useful to run the tests on all PRs, even if they target branches other than `main`

stephprince replied:
Got it, that makes sense. We can remove the branch section entirely then or yes just leave it commented out.

oruebel replied:
Either way is fine. I just commented them out here, because I wanted to see what tests are working/failing on this PR. We can change workflows also in a separate PR and remove this change here before merging if you prefer.

oruebel commented on src/io/ReadIO.hpp:
```diff
@@ -0,0 +1,370 @@
+/**
+ * @brief Generic structure to hold type-erased data and shape
+ */
+class DataBlockGeneric
```
I like the idea of only using `DataBlock` but we'll need to think a bit more about how to do that. One key difference is that `DataBlockGeneric` stores the entire data as `std::any` so we can use a single cast ``std::any_cast<std::vector<DTYPE>>(genericData.data)`` to transform the data to the correct type without having to copy the data. `DataBlock` on the other hand stores the data as and ``std::vector`` and to cast ``std::vector<std::any>`` required copying the data vector and casting each individual element. Similarly, on read we would probably also need to do another copy in order to move the data into the  ``std::vector<std::any>``.  Happy to chat about how we may be able to simplify this. 

Maybe one approach could be to have DataBlock store the data internally as ``std::any`` and then access the data via a function instead so that we do the casting when accessing the data. However, when reading data where the user may not know the data type before-hand, the user still needs to then define the type to cast to since we cannot simply cast `std::any` to `std::vector<std::any>`, but we need to cast to the correct type.

oruebel replied:
I filed #132 to continue this discussion in a separate issue

oruebel commented on src/io/ReadIO.hpp:
```diff
@@ -0,0 +1,370 @@
+   *  StorageObjectType::Attribute but not for other types, e.g., Group or
+   * Undefined.
+   */
+  static_assert(is_allowed_storage_object_type<OTYPE>::value,
```
```suggestion
  static_assert(isAllowedStorageObjectType<OTYPE>::value,
```

oruebel commented on src/io/ReadIO.hpp:
```diff
@@ -0,0 +1,370 @@
+   * select functions for attributes, to not support slicing.
+   */
+  template<StorageObjectType U>
+  struct is_dataset
```
Done.

oruebel commented on src/io/ReadIO.hpp:
```diff
@@ -0,0 +1,370 @@
+  /**
+   * @brief Deleted copy constructor to prevent construction-copying.
+   */
+  ReadDataWrapper(const ReadDataWrapper&) = delete;
```
To be honest, I was reading the term "data" as a generic term for anything that represents data. But I can see how a user may misinterpret "data" as equal to "dataset", so probably best to change it. 

How about instead of ``wrapper`` or `data`, we could use the term `DataObject` to refer to datasets/attributes in a more generic way. I.e., rename to `ReadDataObject` and `BaseRecordingDataObject`

If we want to keep the term ``wrapper`` then renaming``ReadDataWrapper` to `ReadWrapper` seems fine.  However, `BaseRecordingWrapper` may be confusing, because it sounds like it is wrapping a whole recording instead of a single dataset that is used for recording. If `BaseRecordingData` is only used for datasets (and not attributes), then I think calling it `BaseRecordingDataset` or `BaseRecordingDatasetWrapper` would work.

oruebel commented on src/nwb/NWBFile.cpp:
```diff
@@ -143,25 +136,32 @@ Status NWBFile::createElectricalSeries(
+    }
+
+    // Setup electrical series datasets
+    if (!m_io->objectExists(devicePath)) {
```
Good catch. Probably a mistake when I was resolving merge errors. Fixed now.

oruebel commented on src/nwb/base/TimeSeries.hpp:
```diff
@@ -65,79 +44,79 @@ class TimeSeries : public Container
+  template<typename VTYPE = std::any>
+  inline std::unique_ptr<
+      IO::ReadDataWrapper<AQNWB::Types::StorageObjectType::Dataset, VTYPE>>
+  dataLazy() const
```
In #91 I replaced `dataLazy` (and other read fields) with the `DEFINE_FIELD` macro to create these access methods. I changed the name of the functions there to use the naming convention `read<FieldPath>`, e.g., `readData` or `readTimestamps` . To avoid confusion for attributes I used the full path in naming the field, e.g., `readDataUnit` and `readTimestampsUnit`.

I don't think we want to store the read wrappers on the container. They are lightweight and a user may need to use a data type different than the default, so creating the read wrapper on request is useful. I.e., I don't think it is strictly necessary for `BaseRecordingData` and `ReadDataWrapper` to have a common base. We could rename `TimeSeries.data` to `TimeSeries.recordData` to make them more consistent, or we could leave it as is and say that attributes without a prefix are for acquisition and with the prefix are for read only.

With all of this said, I think it is worth discussing whether having a common base class for BaseRecordingData and ReadDataWrapper would be useful.

oruebel commented on src/nwb/hdmf/base/Data.hpp:
```diff
@@ -40,6 +40,7 @@ class Data
 
-  std::unique_ptr<BaseRecordingData> m_dataset;
+  std::unique_ptr<IO::BaseRecordingData>
+      m_dataset;  // TODO We may not want this here if we need Data for read
```
This is same issue as above with `TimeSeries.data` . I removed the ToDo item.

oruebel commented on src/nwb/hdmf/table/DynamicTable.hpp:
```diff
@@ -23,12 +23,13 @@ class DynamicTable : public Container
+  DynamicTable(
+      const std::string& path,
+      std::shared_ptr<IO::BaseIO> io,
+      const std::vector<std::string>& colNames =
+          {});  // TODO Need to remove colNames here and move it to initialize
```
This is still ToDo. We should chat about how we want to deal with colNames. Currently, creating a DynamicTable with empty colNames and then adding columns does not work, but we need to set the colNames at the beginning.

oruebel commented on tests/examples/test_ecephys_data_read.cpp:
```diff
@@ -0,0 +1,215 @@
+    // Get a ReadDatasetWrapper<float> for lazy reading of ElectricalSeries.data
+    // By specifying the value type as a template parameter allows us to read
+    // typed data
+    auto readDataWrapper = electricalSeries->dataLazy<float>();
+    // [example_read_get_data_wrapper_snippet]
+
+    // [example_read_get_datablock_snippet]
+    // Read the full  ElectricalSeries.data back
+    DataBlock<float> dataValues = readDataWrapper->values();
```
Are you refering to 1) ``electricalSeries->dataLazy<float>();`` or to 2) ``DataBlock<float> dataValues = readDataWrapper->values()``. For 1) we could set the default for ``readData`` in #91 to be float in `ElectricalSeries` so we don't need to provide the template. For 2) we could just say `auto dataValues` instead. 

Note, reading with a data type of std::any I don't think will work right now, because we can't simply cast ``std::vector<float>`` to  ``std::vector<std::any>``. I.e., when we don't specify the data type, we can only read as `DataBlockGeneric` but not as `DataBlock` right now.

stephprince replied:
> For 1) we could set the default for readData in https://github.com/NeurodataWithoutBorders/aqnwb/pull/91 to be float in ElectricalSeries so we don't need to provide the template.

I was referring to `electricalSeries->dataLazy<float>();`.  I didn't connect that if we want to read with data type `std::any` then we would have to use `DataBlockGeneric`. 

> we could set the default for readData in https://github.com/NeurodataWithoutBorders/aqnwb/pull/91 to be float in ElectricalSeries so we don't need to provide the template

That would be helpful. It wasn't a huge block to understanding, but I think in general having exactly how the `ReadDataWrapper` is templated influencing the use of `DataBlock` or `DataBlockGeneric` might get tricky. We could discuss this related to the proposal to only use `DataBlock`

oruebel commented on tests/testHDF5IO.cpp:
```diff
@@ -408,11 +415,59 @@ TEST_CASE("SWMRmode", "[hdf5io]")
+{
+  std::vector<int32_t> testData = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+
+  SECTION("read 1D data block of a 1D dataset")
```
The unit tests are still very minimal right now. I totally agree, there are a lot more that need to be added. However, I wanted to get your take on the overall design first before diving even deeper into testing all the different cases.

oruebel commented:
I'm a bit confused why the tests are failing after submitting the suggestions from the review. The build in the test fails with ``/Users/runner/work/aqnwb/aqnwb/src/nwb/ecephys/ElectricalSeries.cpp:61:25: error: use of undeclared identifier 'BaseDataType'; did you mean 'IO::BaseDataType'?
  m_io->createAttribute(BaseDataType::I32,`` But that is not what in line 61 in this PR

https://github.com/NeurodataWithoutBorders/aqnwb/actions/runs/11023552242/job/30615439516?pr=85#step:6:249

oruebel commented:
@stephprince looking at the workflow details https://github.com/NeurodataWithoutBorders/aqnwb/actions/runs/11023552242/job/31901884513?pr=85 The workflow is checking out:

/opt/homebrew/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +b2fb27fdba42dd5183bce7d2cc44421c50e2f428:refs/remotes/pull/85/merge

However, looking at that commit hash at https://github.com/NeurodataWithoutBorders/aqnwb/commit/b2fb27fdba42dd5183bce7d2cc44421c50e2f428 says that _"This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository."_  

Digging around a bit more with ChatGPT, 

_When a GitHub Action runs for a pull request, it often checks out a temporary merge commit. This commit represents the state of the code if the pull request were to be merged into the target branch (usually main). This allows the CI to test the combined code without actually merging it. [...]  The commit b2fb27fdba42dd5183bce7d2cc44421c50e2f428 is likely a temporary merge commit created by GitHub to test the pull request. These commits are not part of the repository's history and exist only in the context of the CI job._

I didn't know that that is what GitHub is doing, but I guess that makes sense that GitHub wants to simulate the state of the code as if the pull request were merged into the target branch. Looking at the code in https://github.com/NeurodataWithoutBorders/aqnwb/commit/b2fb27fdba42dd5183bce7d2cc44421c50e2f428  it indeed shows the error that the build is showing in line `ElectricalSeries.cpp:61:25: error: use of undeclared identifier 'BaseDataType'; did you mean 'IO::BaseDataType'? m_io->createAttribute(BaseDataType::I32, ` I.e., something seems to be going wrong when git tries to automatically merge the branches. 

We could modify the ``tests.yml`` to run all the tests with both states, i.e., with the code as is in the PR as well as with the temporary merge. This would make the CI runtime longer (since all tests would run twice) and make the workflow a bit longer but would help with finding merge errors. So its a tradeoff between runtime and complexity and being more rigorous. What do you think? Here is how the modified ``tests.yml`` workflow could look, where:

- ``test-pr-state`` Job: Checks out the specific PR branch state using ``github.event.pull_request.head.sha``.
-   ``test-merged-state``Job: Checks out the temporary merge commit using ``refs/pull/${{ github.event.pull_request.number }}/merge``.
-    ``sanitize`` Job: Runs after both ``test-pr-state`` and ``test-merged-state jobs``
-    ``validate`` Job: Runs after all previous jobs and performs validation.


```
name: Run tests

on:
  push:
    branches:
      - main
  pull_request:
  workflow_dispatch:

jobs:
  test-pr-state:  # Job to test the state of the PR branch as it is
    if: github.event_name == 'pull_request'  # Run only on pull request events
    name: Test PR State
    defaults:
      run:
        shell: bash
    concurrency:
      group: ${{ github.workflow }}-${{ github.ref }}-${{ matrix.os }}
      cancel-in-progress: true
    strategy:
      fail-fast: false
      matrix:
        os: [macos-latest, ubuntu-latest]

    runs-on: ${{ matrix.os }}

    steps:
      - name: Checkout PR branch
        uses: actions/checkout@v4
        with:
          ref: ${{ github.event.pull_request.head.sha }}  # Checkout the PR branch directly

      - name: Install dependencies - ubuntu
        if: matrix.os == 'ubuntu-latest'
        run: |
          sudo apt-get update
          sudo apt-get install -y libhdf5-dev libboost-all-dev
          git clone https://github.com/catchorg/Catch2.git
          cd Catch2
          git checkout "v3.5.3"
          cmake -Bbuild -H. -DBUILD_TESTING=OFF
          sudo cmake --build build/ --target install

      - name: Install dependencies - macos
        if: matrix.os == 'macos-latest'
        run: brew install hdf5 boost catch2

      - name: Configure
        shell: pwsh
        run: cmake "--preset=ci-$("${{ matrix.os }}".split("-")[0])"

      - name: Build
        run: cmake --build build --config Release -j 2

      - name: Install
        run: cmake --install build --config Release --prefix prefix

      - name: Test
        working-directory: build
        run: ctest --output-on-failure --no-tests=error -C Release -j 2

      - name: Upload artifacts
        uses: actions/upload-artifact@v3
        with:
          name: test-files-${{ matrix.os }}
          path: |
            build/tests/data/*.nwb

  test-merged-state:  # Job to test the state of the code after merging the PR into the target branch
    if: github.event_name == 'pull_request'  # Run only on pull request events
    name: Test Merged State
    defaults:
      run:
        shell: bash
    concurrency:
      group: ${{ github.workflow }}-${{ github.ref }}-${{ matrix.os }}
      cancel-in-progress: true
    strategy:
      fail-fast: false
      matrix:
        os: [macos-latest, ubuntu-latest]

    runs-on: ${{ matrix.os }}

    steps:
      - name: Checkout merged state
        uses: actions/checkout@v4
        with:
          ref: refs/pull/${{ github.event.pull_request.number }}/merge  # Checkout the merge commit

      - name: Install dependencies - ubuntu
        if: matrix.os == 'ubuntu-latest'
        run: |
          sudo apt-get update
          sudo apt-get install -y libhdf5-dev libboost-all-dev
          git clone https://github.com/catchorg/Catch2.git
          cd Catch2
          git checkout "v3.5.3"
          cmake -Bbuild -H. -DBUILD_TESTING=OFF
          sudo cmake --build build/ --target install

      - name: Install dependencies - macos
        if: matrix.os == 'macos-latest'
        run: brew install hdf5 boost catch2

      - name: Configure
        shell: pwsh
        run: cmake "--preset=ci-$("${{ matrix.os }}".split("-")[0])"

      - name: Build
        run: cmake --build build --config Release -j 2

      - name: Install
        run: cmake --install build --config Release --prefix prefix

      - name: Test
        working-directory: build
        run: ctest --output-on-failure --no-tests=error -C Release -j 2

      - name: Upload artifacts
        uses: actions/upload-artifact@v3
        with:
          name: test-files-${{ matrix.os }}
          path: |
            build/tests/data/*.nwb

  sanitize:  # Job to run sanitization tests on the merged state
    if: github.event_name == 'pull_request'  # Run only on pull request events
    needs: [test-pr-state, test-merged-state]

    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4
        with:
          ref: refs/pull/${{ github.event.pull_request.number }}/merge  # Checkout the merge commit

      - name: Install dependencies
        run: |
          sudo apt-get update
          sudo apt-get install -y libhdf5-dev libboost-all-dev
          git clone https://github.com/catchorg/Catch2.git
          cd Catch2
          git checkout "v3.5.3"
          cmake -Bbuild -H. -DBUILD_TESTING=OFF
          sudo cmake --build build/ --target install

      - name: Configure
        run: cmake --preset=ci-sanitize

      - name: Build
        run: cmake --build build/sanitize -j 2

      - name: Test
        working-directory: build/sanitize
        env:
          ASAN_OPTIONS: "strict_string_checks=1:\
            detect_stack_use_after_return=1:\
            check_initialization_order=1:\
            strict_init_order=1:\
            detect_leaks=1:\
            halt_on_error=1"
          UBSAN_OPTIONS: "print_stacktrace=1:\
            halt_on_error=1"
        run: ctest --output-on-failure --no-tests=error -j 2

  validate:  # Job to validate the results using the test files
    needs: [test-pr-state, test-merged-state, sanitize]
    defaults:
      run:
        shell: bash
    concurrency:
      group: ${{ github.workflow }}-${{ github.ref }}-${{ matrix.os }}
      cancel-in-progress: true
    strategy:
      fail-fast: false
      matrix:
        os: [macos-latest, ubuntu-latest]

    runs-on: ${{ matrix.os }}

    steps:
      - name: Download test files
        uses: actions/download-artifact@v3
        with:
          name: test-files-${{ matrix.os }}
          path: nwb_files

      - name: Set up Python
        uses: actions/setup-python@v5
        with:
          python-version: '3.12'

      - name: Install pynwb and run validation
        run: |
          python -m pip install --upgrade pip
          python -m pip install nwbinspector
          nwbinspector nwb_files --threshold BEST_PRACTICE_VIOLATION
```

oruebel commented:
> We could modify the `tests.yml` to run all the tests with both states, i.e., with the code as is in the PR as well as with the temporary merge. This would make the CI runtime longer (since all tests would run twice) and make the workflow a bit longer but would help with finding merge errors.

We decide not to do this and to continue testing only for the merged version. We decided to add a note in the developer docs to clarify this behavior.

oruebel commented on src/nwb/RegisteredType.hpp:
```diff
@@ -0,0 +1,327 @@
+   * description \
+   */ \
+  template<typename VTYPE = default_type> \
+  inline std::unique_ptr<IO::ReadDataWrapper<storageObjectType, VTYPE>> name() \
```
```suggestion
  inline virtual std::unique_ptr<IO::ReadDataWrapper<storageObjectType, VTYPE>> name() \
```

We should add the ``virtual`` keyword here to ensure when overriding with `DEFINE_FIELD` in subclasses, that the function resolution works as expected. I am not sure how much sense it makes to keep `inline` as well or whether that should be removed.

oruebel replied:
Actually, that is not going to work, since the function is also a template function and virtual and templates don't like each other. As it is currently, when`DEFINE_FIELD` is specified again in a derived class for the same field, then the function will be hidden instead of overriden . I.e., if we call the function on the derived class everything works as expected. However, if we cast the Derived pointer to the base class, then we will call the function as defined on the base class, because it is not virtual. I think in most cases this should not be a problem, but is something we should be aware of, and is worth documenting. 

A possible workaround would be to have `DEFINE_FIELD` generate two member functions. A templated function (as it is right now) and then a second virtual function that calls the templated function with the `default_type`

```
 virtual std::unique_ptr<IO::ReadDataWrapper<storageObjectType, default_type>> name##_virtual() const \
  { \
    return name<default_type>(); \
  }
```

However, I am not sure this is necessary and I think could lead to more confusion. Maybe we can just make an issue for this for now just have a record for this.

oruebel replied:
I created #117 for this. I think it is fine to address this later. For now, I added the following note to the docs:

https://github.com/NeurodataWithoutBorders/aqnwb/blob/7b88122ce7a1855f5c640d8235cfe18ab3355c01/docs/pages/devdocs/registered_types.dox#L82-L87

oruebel commented:
@stephprince I synced the branch with the main branch. However, Windows tests are currently failing due to ``'boost/multi_array.hpp': No such file or directory``. Can you check the Windows action to make sure boost is being installed correctly.

oruebel commented:
> , Windows tests are currently failing due to `'boost/multi_array.hpp': No such file or directory`

I fixed the include error for `boost/multi_array.hpp` but it looks like there are a few more errors on Windows that need to be fixed.

oruebel commented:
@stephprince Looking through the remaining ToDo items, I think we can reasonably create separate issues and PRs for those. I think this is now reasonably well tested that we can merge this and create new issues for the missing items.

stephprince reviewed (approved):
That sounds good to me! :tada:

oruebel commented:
> That sounds good to me! 🎉

👍 thanks for approving the PR. Here the list of follow-up issues:

# Items moved to new issues
* #115
* #114 
* #113
* #127
* #129 and also related to #107
* #130
* #131
* #132
* #133
* #134
